exporter/exporterhelper/persistent_queue.go

```diff
@@ -0,0 +1,343 @@
+
+	readIndexBytes, err := wcs.client.Get(ctx, wcs.readIndexKey)
+	if err != nil || readIndexBytes == nil {
+		wcs.logger.Debug("failed getting read index, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+		wcs.readIndex = 0
+	} else {
+		val, conversionErr := bytesToUint64(readIndexBytes)
+		if conversionErr != nil {
+			wcs.logger.Warn("read index corrupted, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+			wcs.readIndex = 0
+		} else {
+			wcs.readIndex = val
+		}
+	}
+
+	writeIndexBytes, err := wcs.client.Get(ctx, wcs.writeIndexKey)
+	if err != nil || writeIndexBytes == nil {
+		wcs.logger.Debug("failed getting write index, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+		wcs.writeIndex = 0
+	} else {
+		val, conversionErr := bytesToUint64(writeIndexBytes)
+		if conversionErr != nil {
+			wcs.logger.Warn("write index corrupted, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+			wcs.writeIndex = 0
+		} else {
+			wcs.writeIndex = val
+		}
+	}
+}
+
+// Put marshals the request and puts it into the persistent queue
+func (pcs *persistentContiguousStorage) put(req request) error {
+	buf, err := req.marshal()
+	if err != nil {
+		return err
+	}
+
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	itemKey := pcs.buildItemKey(pcs.writeIndex)
+	err = pcs.client.Set(context.Background(), itemKey, buf)
+	if err != nil {
+		return err
+	}
+
+	pcs.writeIndex++
+	writeIndexBytes, err := uint64ToBytes(pcs.writeIndex)
+	if err != nil {
+		pcs.logger.Warn("failed converting write index uint64 to bytes", zap.Error(err))
+	}
+
+	return pcs.client.Set(context.Background(), pcs.writeIndexKey, writeIndexBytes)
+}
+
+// getNextItem pulls the next available item from the persistent storage; if none is found, returns (nil, false)
+func (pcs *persistentContiguousStorage) getNextItem(ctx context.Context) (request, bool) {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	if pcs.readIndex < pcs.writeIndex {
+		itemKey := pcs.buildItemKey(pcs.readIndex)
+		// Increase here, so despite errors it would still progress
+		pcs.readIndex++
+
+		buf, err := pcs.client.Get(ctx, itemKey)
+		if err != nil {
+			pcs.logger.Error("error when getting item from persistent storage",
+				zap.String(queueNameKey, pcs.queueName), zap.String(zapItemKey, itemKey), zap.Error(err))
+			return nil, false
+		}
+		req, unmarshalErr := pcs.unmarshaler(buf)
+		pcs.updateItemRead(ctx, itemKey)
+		if unmarshalErr != nil {
+			pcs.logger.Error("error when unmarshalling item from persistent storage",
+				zap.String(queueNameKey, pcs.queueName), zap.String(zapItemKey, itemKey), zap.Error(unmarshalErr))
+			return nil, false
+		}
+
+		return req, true
+	}
+
+	return nil, false
+}
+
+func (pcs *persistentContiguousStorage) loop(ctx context.Context) {
+	for {
+		if pcs.stopped.Load() != 0 {
+			return
+		}
+
+		req, found := pcs.getNextItem(ctx)
+		if found {
+			pcs.reqChan <- req
+		} else {
+			time.Sleep(pcs.retryDelay)
+		}
+	}
+}
```

Comment:
```suggestion
func (pcs *persistentContiguousStorage) loop() {
	for {
		if pcs.stopped.Load() != 0 {
			return
		}

		req, found := pcs.getNextItem(context.Background())
		if found {
			pcs.reqChan <- req
		} else {
			time.Sleep(pcs.retryDelay)
		}
	}
}
```

Possibly add a timeout to the context passed into `getNextItem`?

If you use a channel instead of `stopped` atomic int then you''ll be able to asynchronously wait on that channel (for closure) and the `pcs.reqChan <- req` branch hence not blocking in case someone stops `persistentContiguousStorage` when you try to send to `reqChan`.

Reply:
I am not sure if I understand how that would work. Can you share an example?

Reply:
Assuming there's a `persistentContiguousStorage.stopChan` (closing of which which will make is continuously send channel's default value)

```go
func (pcs *persistentContiguousStorage) loop() {
	for {
		req, found := pcs.getNextItem(context.Background())
		if found {
			select {
			case <-pcs.stopChan:
				return
			case pcs.reqChan <- req:
			}
		} else {
			select {
			case <-pcs.stopChan:
				return
			case <-time.After(pcs.retryDelay):
			}
		}
	}
}
```

Reply:
This is beautiful, thank you @pmalek-sumo !